pkg/randchar/generator_test.go

```diff
@@ -0,0 +1,430 @@
+			n:                 3,
+		},
+		"multiple minima": {
+			base:   Alphanumeric,
+			minima: []minimum{{1, Numeric}, {1, Uppercase}},
+			n:      3,
+		},
+		"multiple minima == n": {
+			base:   Alphanumeric,
+			minima: []minimum{{2, Numeric}, {1, Uppercase}},
+			n:      3,
+		},
+		"multiple minima > n": {
+			base:                  Alphanumeric,
+			minima:                []minimum{{2, Numeric}, {2, Uppercase}},
+			n:                     3,
+			expectedGenerateError: errors.New("n cannot be smaller than the minimum required length of the generator"),
+		},
+		"multiple minima with same charset small first": {
+			base:   Alphanumeric,
+			minima: []minimum{{1, Numeric}, {2, Numeric}},
+			n:      3,
+		},
+		"multiple minima with same charset big first": {
+			base:   Alphanumeric,
+			minima: []minimum{{2, Numeric}, {1, Numeric}},
+			n:      3,
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			options := make([]Option, len(tc.minima))
+			for i, min := range tc.minima {
+				options[i] = Min(min.count, min.charset)
+			}
+
+			rand, err := NewRand(tc.base, options...)
+			assert.Equal(t, err, tc.expectedInitError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			actual, err := rand.Generate(tc.n)
+			assert.Equal(t, err, tc.expectedGenerateError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			for _, min := range tc.minima {
+				count := countFromSet(actual, min.charset)
+				if count < min.count {
+					t.Errorf("unexpected count for %v: %d (actual) < %d (expected minimum)", min.charset, count, min.count)
+				}
+			}
+		})
+	}
+}
+
+func TestMultipleMin(t *testing.T) {
+	cases := map[string]struct {
+		base                  Charset
+		minima                []minimum
+		expectedInitError     error
+		n                     int
+		expectedGenerateError error
+	}{}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			options := make([]Option, len(tc.minima))
+			for i, min := range tc.minima {
+				options[i] = Min(min.count, min.charset)
+			}
+
+			rand, err := NewRand(tc.base, options...)
+			assert.Equal(t, err, tc.expectedInitError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			actual, err := rand.Generate(tc.n)
+			assert.Equal(t, err, tc.expectedGenerateError)
+
+			// skip other assertions if error case
+			if err != nil {
+				return
+			}
+
+			for _, min := range tc.minima {
+				count := countFromSet(actual, min.charset)
+				if count < min.count {
+					t.Errorf("unexpected count for %v: %d (actual) < %d (expected minimum)", min.charset, count, min.count)
+				}
+			}
+		})
+	}
+
+}
+
+func TestNewCharset(t *testing.T) {
+	cases := map[string]struct {
+		in       string
+		expected string
+	}{
+		"empty": {
+			in:       "",
+			expected: "",
+		},
+		"duplicates": {
+			in:       "aa",
+			expected: "a",
+		},
+		"multiple": {
```

Comment:
What are we testing here?

`NewCharset("abc") == Charset("abc")`?

I think it would make more sense to test the underlying byte array in this unit-test and test the equals thing in the Equals unit test (already exists: https://github.com/secrethub/secrethub-go/pull/81/files#diff-ae889a7c8a8642d31b52d28cf0280c68R341).

Reply:
We're testing whether multiple characters leads to a set with multiple characters. It's not a super interesting test, but we should have it nonetheless IMO. I changed it so the equality test is indeed more a unit test and doesn't depend on the `Charset.Equals` test.